[Remote Inspection] Make the repeated element targeting heuristic robust in the case where elements are unparented

whsieh opened the pull request:
#### 8af90c08dbd9df112cb58f4ed043180dc8c7d512
<pre>
[Remote Inspection] Make the repeated element targeting heuristic robust in the case where elements are unparented
<a href="https://bugs.webkit.org/show_bug.cgi?id=271848">https://bugs.webkit.org/show_bug.cgi?id=271848</a>

Reviewed by Aditya Keerthi and Abrar Rahman Protyasha.

Currently, client rects for visibility adjustment are added at visibility adjustment time, based on
the current client rect of the target element. This has a major drawback, in that a target could be
unparented (or otherwise prematurely hidden) by the page, and we&apos;ll fail to add any client rects
corresponding to the target element.

Instead, this patch refactors that mechanism to be more robust, by computing and caching the last
set of targeted element client rects up front when finding targets. Later, when we actually go and
apply visibility adjustment, we&apos;ll consult these cached target rects instead of reading the current
element state.

Test: ElementTargeting.AdjustVisibilityForUnparentedElement

* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::computeClientRect):

Move this static helper function up in this file.

(WebCore::ElementTargetingController::findTargets):
(WebCore::ElementTargetingController::adjustVisibility):

Also refactor this logic to update adjustment regions, such that it happens before any of the
targeted elements are hidden. This is possible now that we only rely on the element identifiers and
`m_pendingAdjustmentClientRects` (from the last element targeting call).

(WebCore::ElementTargetingController::resetAdjustmentRegions):
* Source/WebCore/page/ElementTargetingController.h:

Add `m_pendingAdjustmentClientRects`, which tracks the last set of targeted elements which might be
imminently hidden. Keeping around only this recent state helps avoided unbounded growth as the user
targets more elements around the page, but still allows us to handle this corner case where targeted
elements can no longer be hidden.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::adjustVisibilityForTargetedElements):

Adjust the WebCore method to take a list of element identifiers instead of elements; see above for
more details.

* Tools/TestWebKitAPI/PlatformUtilities.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
(-[WKWebView adjustVisibilityForTargets:]):

Add a new API test to exercise the change.

(TestWebKitAPI::TEST(ElementTargeting, AdjustVisibilityForUnparentedElement)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/GPUProcess.mm:
(TestWebKitAPI::TEST(GPUProcess, CanvasBasicCrashHandling)):
(TestWebKitAPI::convertToCGImage): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewSnapshot.mm:
(-[TestSnapshotWrapper takeSnapshotWithWebView:configuration:completionHandler:]):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageError)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageEmptyRect)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageZeroWidth)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageZeroSizeView)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageZeroSizeViewNoConfiguration)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageBaseCase)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageScale)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageNilConfiguration)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageUninitializedConfiguration)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageUninitializedSnapshotWidth)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageLargeAsyncDecoding)):
(TestWebKitAPI::TEST(WKWebView, SnapshotAfterScreenUpdates)):
(TestWebKitAPI::TEST(WKWebView, SnapshotWithoutAfterScreenUpdates)):
(TestWebKitAPI::TEST(WKWebView, SnapshotWebGL)):
(TestWebKitAPI::TEST(WKWebView, SnapshotWithoutSelectionHighlighting)):
(TestWebKitAPI::TEST(WKWebView, SnapshotWithContentsRect)):

Refactor some existing API tests to use the new `CGImage` conversion helpers.

(convertToCGImage): Deleted.
(TEST(WKWebView, SnapshotImageError)): Deleted.
(TEST(WKWebView, SnapshotImageEmptyRect)): Deleted.
(TEST(WKWebView, SnapshotImageZeroWidth)): Deleted.
(TEST(WKWebView, SnapshotImageZeroSizeView)): Deleted.
(TEST(WKWebView, SnapshotImageZeroSizeViewNoConfiguration)): Deleted.
(TEST(WKWebView, SnapshotImageEmptyWithOutOfScopeCompletionHandler)): Deleted.
(TEST(WKWebView, SnapshotImageBaseCase)): Deleted.
(TEST(WKWebView, SnapshotImageScale)): Deleted.
(TEST(WKWebView, SnapshotImageNilConfiguration)): Deleted.
(TEST(WKWebView, SnapshotImageUninitializedConfiguration)): Deleted.
(TEST(WKWebView, SnapshotImageUninitializedSnapshotWidth)): Deleted.
(TEST(WKWebView, SnapshotImageLargeAsyncDecoding)): Deleted.
(TEST(WKWebView, SnapshotAfterScreenUpdates)): Deleted.
(TEST(WKWebView, SnapshotWithoutAfterScreenUpdates)): Deleted.
(TEST(WKWebView, SnapshotWebGL)): Deleted.
(TEST(WKWebView, SnapshotWithoutSelectionHighlighting)): Deleted.
(TEST(WKWebView, SnapshotWithContentsRect)): Deleted.

Refactor some existing API tests to use the new `CGImage` conversion helpers.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/element-targeting-2.html:
* Tools/TestWebKitAPI/cocoa/PlatformUtilitiesCocoa.mm:
(TestWebKitAPI::Util::convertToCGImage):

Pull out some helper functions to convert platform `NSImage` / `UIImage` into `CGImage`.

* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView snapshotAfterScreenUpdates]):

Add a new helper method for testing, that grabs a snapshot of the web view (after screen updates) as
a `CGImage`.

Canonical link: <a href="https://commits.webkit.org/276824@main">https://commits.webkit.org/276824@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/12e0a08c22abf36a903cf09e7f6509c6082aa804

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45802 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24930 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48385 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48472 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41839 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48108 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29217 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22328 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37495 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46380 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22008 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39507 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18663 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/19407 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40602 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3846 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42108 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40951 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50232 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20794 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17301 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/44614 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22098 "Built successfully") | | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43481 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state-changed, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10169 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22457 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21788 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
```diff
@@ -153,4 +167,39 @@ @implementation _WKTargetedElementInfo (TestingAdditions)
+#if PLATFORM(IOS_FAMILY)
+    auto window = adoptNS([[UIWindow alloc] initWithFrame:[webView frame]]);
+    [window addSubview:webView.get()];
+#endif
```
I think you can use `-[TestWKWebView initWithFrame:configuration:addToWindow:]` rather than creating a window here.

whsieh replied:
Good catch!

aprotyas commented on Tools/TestWebKitAPI/PlatformUtilities.h:
```diff
@@ -40,6 +40,14 @@ OBJC_CLASS NSDictionary;
+OBJC_CLASS NSWindow;
+OBJC_CLASS UIImage;
+OBJC_CLASS UIWindow;
+typedef struct CGImage *CGImageRef;
```
Not that it matters, but optionally:
```suggestion
using CGImageRef = CGImage*;
```

whsieh replied:
Will do!

whsieh committed:
[Remote Inspection] Make the repeated element targeting heuristic robust in the case where elements are unparented
https://bugs.webkit.org/show_bug.cgi?id=271848

Reviewed by Aditya Keerthi and Abrar Rahman Protyasha.

Currently, client rects for visibility adjustment are added at visibility adjustment time, based on
the current client rect of the target element. This has a major drawback, in that a target could be
unparented (or otherwise prematurely hidden) by the page, and we'll fail to add any client rects
corresponding to the target element.

Instead, this patch refactors that mechanism to be more robust, by computing and caching the last
set of targeted element client rects up front when finding targets. Later, when we actually go and
apply visibility adjustment, we'll consult these cached target rects instead of reading the current
element state.

Test: ElementTargeting.AdjustVisibilityForUnparentedElement

* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::computeClientRect):

Move this static helper function up in this file.

(WebCore::ElementTargetingController::findTargets):
(WebCore::ElementTargetingController::adjustVisibility):

Also refactor this logic to update adjustment regions, such that it happens before any of the
targeted elements are hidden. This is possible now that we only rely on the element identifiers and
`m_pendingAdjustmentClientRects` (from the last element targeting call).

(WebCore::ElementTargetingController::resetAdjustmentRegions):
* Source/WebCore/page/ElementTargetingController.h:

Add `m_pendingAdjustmentClientRects`, which tracks the last set of targeted elements which might be
imminently hidden. Keeping around only this recent state helps avoided unbounded growth as the user
targets more elements around the page, but still allows us to handle this corner case where targeted
elements can no longer be hidden.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::adjustVisibilityForTargetedElements):

Adjust the WebCore method to take a list of element identifiers instead of elements; see above for
more details.

* Tools/TestWebKitAPI/PlatformUtilities.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
(-[WKWebView adjustVisibilityForTargets:]):

Add a new API test to exercise the change.

(TestWebKitAPI::TEST(ElementTargeting, AdjustVisibilityForUnparentedElement)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/GPUProcess.mm:
(TestWebKitAPI::TEST(GPUProcess, CanvasBasicCrashHandling)):
(TestWebKitAPI::convertToCGImage): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewSnapshot.mm:
(-[TestSnapshotWrapper takeSnapshotWithWebView:configuration:completionHandler:]):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageError)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageEmptyRect)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageZeroWidth)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageZeroSizeView)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageZeroSizeViewNoConfiguration)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageBaseCase)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageScale)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageNilConfiguration)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageUninitializedConfiguration)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageUninitializedSnapshotWidth)):
(TestWebKitAPI::TEST(WKWebView, SnapshotImageLargeAsyncDecoding)):
(TestWebKitAPI::TEST(WKWebView, SnapshotAfterScreenUpdates)):
(TestWebKitAPI::TEST(WKWebView, SnapshotWithoutAfterScreenUpdates)):
(TestWebKitAPI::TEST(WKWebView, SnapshotWebGL)):
(TestWebKitAPI::TEST(WKWebView, SnapshotWithoutSelectionHighlighting)):
(TestWebKitAPI::TEST(WKWebView, SnapshotWithContentsRect)):

Refactor some existing API tests to use the new `CGImage` conversion helpers.

(convertToCGImage): Deleted.
(TEST(WKWebView, SnapshotImageError)): Deleted.
(TEST(WKWebView, SnapshotImageEmptyRect)): Deleted.
(TEST(WKWebView, SnapshotImageZeroWidth)): Deleted.
(TEST(WKWebView, SnapshotImageZeroSizeView)): Deleted.
(TEST(WKWebView, SnapshotImageZeroSizeViewNoConfiguration)): Deleted.
(TEST(WKWebView, SnapshotImageEmptyWithOutOfScopeCompletionHandler)): Deleted.
(TEST(WKWebView, SnapshotImageBaseCase)): Deleted.
(TEST(WKWebView, SnapshotImageScale)): Deleted.
(TEST(WKWebView, SnapshotImageNilConfiguration)): Deleted.
(TEST(WKWebView, SnapshotImageUninitializedConfiguration)): Deleted.
(TEST(WKWebView, SnapshotImageUninitializedSnapshotWidth)): Deleted.
(TEST(WKWebView, SnapshotImageLargeAsyncDecoding)): Deleted.
(TEST(WKWebView, SnapshotAfterScreenUpdates)): Deleted.
(TEST(WKWebView, SnapshotWithoutAfterScreenUpdates)): Deleted.
(TEST(WKWebView, SnapshotWebGL)): Deleted.
(TEST(WKWebView, SnapshotWithoutSelectionHighlighting)): Deleted.
(TEST(WKWebView, SnapshotWithContentsRect)): Deleted.

Refactor some existing API tests to use the new `CGImage` conversion helpers.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/element-targeting-2.html:
* Tools/TestWebKitAPI/cocoa/PlatformUtilitiesCocoa.mm:
(TestWebKitAPI::Util::convertToCGImage):

Pull out some helper functions to convert platform `NSImage` / `UIImage` into `CGImage`.

* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView snapshotAfterScreenUpdates]):

Add a new helper method for testing, that grabs a snapshot of the web view (after screen updates) as
a `CGImage`.

Canonical link: https://commits.webkit.org/276824@main
